File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def bfs(arr, si, sj, N, M):
    q = deque()
    v = [[0]*M for _ in range(N)]
    
    q.append((si,sj))
    v[si][sj]=1
    
    while q:
        ci,cj = q.popleft()
        if (ci,cj)==(N-1,M-1):
            return v[ci][cj]
        
        # 네방향, 범위내, 미방문, 조건: arr[][]==1
        for di,dj in ((-1,0),(1,0),(0,1),(0,-1)):
            ni,nj = ci+di, cj+dj
            
            if 0<=ni<N and 0<=nj<M and v[ni][nj]==0 and arr[ni][nj]==1:
                q.append((ni,nj))
                v[ni][nj]=v[ci][cj]+1
    return -1
    
def solution(maps):
    answer = 0
    
    M = len(maps[0])
    N = len(maps)
    #print(N,M)
    
    answer = bfs(maps, 0, 0, N, M) # 0-based
    return answer

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    static class Point {
        int x;
        int y;
        int dist;
        
        Point(int x, int y, int dist) {
            this.x = x;
            this.y = y;
            this.dist = dist;
        }
    }
    
    static int[] dx = {-1,1,0,0};
    static int[] dy = {0,0,-1,1};
    
    int bfs(int[][] maps, int N, int M) {
        Queue<Point> q = new LinkedList<>();
        boolean[][] visited = new boolean[N][M];
        
        q.offer(new Point(N-1, M-1, 1));
        visited[N-1][M-1] = true; // 항상 똑같은 위치에서 시작
        
        boolean flag = false;
        
        while (!q.isEmpty()) {
            Point cur = q.poll();
            if (cur.x==0 && cur.y==0) return cur.dist;
            
            flag = false;
            for (int i=0; i<4; i++) {
                int nx = cur.x+dx[i];
                int ny = cur.y+dy[i];
                
                if(nx<0||nx>=N||ny<0||ny>=M) continue;
                if (visited[nx][ny]) continue;
                if (maps[nx][ny]==0) continue;
                
                q.offer(new Point(nx, ny, cur.dist+1));
                visited[nx][ny] = true;
                flag = true;
            }
        } 
        
        return -1;
    }
    
    public int solution(int[][] maps) {
        int answer = 0;
        int N = maps.length;
        int M = maps[0].length;
        
        answer = bfs(maps, N, M);
        
        return answer;
    }
}